Type annotations, part 1

crusaderky opened the pull request:
Fix mypy errors in *most* files; a handful of files remain for later.
Does not enable actual mypy yet.

crusaderky commented:
I've decided to have a first tranche of this work merged. A few files remain that fail mypy, but I'll tackle them later on.
All test failures are already visible in main. Ready for review and merge.

crusaderky commented on dask/array/tests/test_cupy_slicing.py:
```diff
@@ -34,7 +34,7 @@ def test_index_with_int_dask_array(x_chunks, idx_chunks):
 @pytest.mark.parametrize("idx_chunks", [None, 3, 2, 1])
 @pytest.mark.parametrize("x_chunks", [(3, 5), (2, 3), (1, 2), (1, 1)])
-def test_index_with_int_dask_array(x_chunks, idx_chunks):
+def test_index_with_int_dask_array_nep35(x_chunks, idx_chunks):
```
avoid test name conflict

ian-r-rose commented on dask/config.py:
```diff
@@ -526,7 +558,7 @@ def update_defaults(new, config=config, defaults=defaults):
 
 
-def expand_environment_variables(config):
+def expand_environment_variables(config: Any) -> Any:
```
I don't want to break backwards compatibility, so the `Any` branch should stay, but I wish it weren't here.

ian-r-rose commented on dask/tests/test_context.py:
```diff
@@ -36,7 +36,7 @@ def bar():
 class Foo:
     @globalmethod(key="f")
-    def f():
+    def f():  # type: ignore
```
I must admit, I have no idea what the error `mypy` is reporting here is. Seems perfectly fine to me.

crusaderky replied:
The ``@globalmethod`` decorator is too magic for mypy to handle.

ian-r-rose commented on dask/widgets/__init__.py:
```diff
@@ -9,11 +9,11 @@
     TEMPLATE_PATHS = []
 
-    def get_environment(*args, **kwargs):
+    def get_environment(*args, **kwargs):  # type: ignore
```
Doesn't seem like it costs anything to make the signatures of this match that of `widgets.get_{environment, template}`:
```suggestion
    def get_environment() -> Environment:
```
Is anyone using this and passing args/kwargs into the function?

crusaderky replied:
fixed

ian-r-rose commented on dask/widgets/__init__.py:
```diff
@@ -9,11 +9,11 @@
         raise ImportError(msg) from exception
 
-    def get_template(*args, **kwargs):
+    def get_template(*args, **kwargs):  # type: ignore
```
Here as well

ian-r-rose commented on docs/source/develop.rst:
```diff
@@ -263,10 +263,10 @@ Docstring testing requires ``graphviz`` to be installed. This can be done via::
-CI. Developers should run them locally before they submit a PR, through the single
-command ``pre-commit run --all-files``. This makes sure that linter versions and options
-are aligned for all developers.
+Dask uses several code linters (flake8, black, isort, pyupgrade, mypy), which are
```
:tada:

ian-r-rose commented on docs/source/scripts/scheduling.py:
```diff
@@ -60,7 +60,7 @@ def dense(width, height):
         dsk, keys = trivial(int(n), 5)
         start = time()
-        get(dsk, keys)
+        get(dsk, keys)  # type: ignore
```
Sigh, another weird one from `mypy`. Do you understand this? Pyright seems okay with it

crusaderky replied:
I suspect that
- the functions in that list are not typed and/or
- the functions in that list don't have the exact same signature

ian-r-rose replied:
Yeah, for sure it's related to having untyped signatures in the above functions, but my understanding was that `mypy` will typically infer `Any` for such a case, and instead I'm seeing `cannot call function of unknown type`. But not worth going into, I just thought it was weird.

ian-r-rose reviewed (commented):
Thanks @crusaderky, this is awesome. Can't wait to start using it. I have a few questions/suggestions, but nothing that I'd consider blocking.

Can you remind me what the problem was with configuring mypy to ignore problematic files? If we can, I'd love to start turning this on and just exclude some things for the time being.

crusaderky commented on dask/array/tests/test_atop.py:
```diff
@@ -683,7 +683,7 @@ def test_args_delayed():
 
 @pytest.mark.parametrize(
-    "tup", [(1, 2), collections.namedtuple("foo", ["a", "b"])(1, 2)]
+    "tup", [(1, 2), collections.namedtuple("foo", ["a", "b"])(1, 2)]  # type: ignore
```
No, but then again, we're using namedtuple improperly.
This works:
```python
T = collections.namedtuple("T", ["a", "b"])
T(1, 2)
```

ian-r-rose replied:
I don't understand why it's improper to not capture the type, is that documented anywhere? I don't see it [here](https://docs.python.org/3/library/collections.html#collections.namedtuple).

crusaderky commented on dask/callbacks.py:
```diff
@@ -43,7 +46,7 @@ class Callback:
     """
 
-    active = set()
+    active: ClassVar[set[tuple]] = set()
```
done

crusaderky commented:
> Can you remind me what the problem was with configuring mypy to ignore problematic files? 

Anything importing the problematic files will stop working.

> If we can, I'd love to start turning this on and just exclude some things for the time being.

You can add to .pre-commit-config.yaml:
```yaml
  - repo: https://github.com/pre-commit/mirrors-mypy
    rev: v0.931
    hooks:
      - id: mypy
        additional_dependencies:
          # Type stubs
          - types-docutils
          - types-pkg_resources
          - types-PyYAML
          - types-requests
          - types-setuptools
          # Libraries exclusively imported under `if TYPE_CHECKING:`
          - typing_extensions  # To be reviewed after dropping Python 3.7
          # Typed libraries
          - numpy
```
and then run ``pre-commit run --all-files | grep -v <known bad files>``

> Here and elsewhere with the same pattern: it's annoying, but we could remove the ignore by splitting into two lines:

Your proposed alternative hurts readability and is overkill for a unit test, IMHO

I think I addressed all review comments

ian-r-rose commented:
> > Here and elsewhere with the same pattern: it's annoying, but we could remove the ignore by splitting into two lines:
> 
> Your proposed alternative hurts readability and is overkill for a unit test, IMHO

Fair enough. I don't love the proposed solution either, but might advocate for it more strongly if it were not in a unit test.

ian-r-rose reviewed (approved):
Thanks @crusaderky!

jrbourbeau commented on docs/source/develop.rst:
```diff
@@ -263,10 +263,10 @@ Docstring testing requires ``graphviz`` to be installed. This can be done via::
-CI. Developers should run them locally before they submit a PR, through the single
-command ``pre-commit run --all-files``. This makes sure that linter versions and options
-are aligned for all developers.
+Dask uses several code linters (flake8, black, isort, pyupgrade, mypy), which are
```
I could be missing something, but it looks like we're not enforcing `mypy` at the moment. Should we drop `mypy` from this list?

crusaderky replied:
Ah, you're right. That's from the initial version with the checker. Yes this paragraph should be reverted for now.
